From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: kernel: 3215 tty hang
Patch-mainline: not yet
Git-commit: 26d766c60f4ea08cd14f0f3435a6db3d6cc2ae96
References: bnc#891087, LTC#114562

Description:  kernel: 3215 tty hang
Symptom:      System hang while waiting for the flush of the 3215 tty
              output buffer
Problem:      If the start of an 3215 I/O is unsuccessful the driver
              fails to set up a retry timer correctly. This can leave
              the system in a hanging state if the init process waits
              for the completion of the flush of the 3215 tty.
Solution:     Check for queued requests after the start of 3215 I/O
              and set up a timer if output is pending.
Reproduction: Starting a Linux system sometimes hangs in the early boot
              process.

Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John jolly <jjolly@suse.de>
---
 drivers/s390/char/con3215.c |   40 +++++++++++++++++++++-------------------
 1 file changed, 21 insertions(+), 19 deletions(-)

--- a/drivers/s390/char/con3215.c
+++ b/drivers/s390/char/con3215.c
@@ -290,13 +290,17 @@ static void raw3215_timeout(unsigned lon
 	unsigned long flags;
 
 	spin_lock_irqsave(get_ccwdev_lock(raw->cdev), flags);
-	if (raw->flags & RAW3215_TIMER_RUNS) {
-		del_timer(&raw->timer);
-		raw->flags &= ~RAW3215_TIMER_RUNS;
-		if (!(raw->flags & RAW3215_FROZEN)) {
-			raw3215_mk_write_req(raw);
-			raw3215_start_io(raw);
-		}
+	raw->flags &= ~RAW3215_TIMER_RUNS;
+	if (!(raw->flags & RAW3215_FROZEN)) {
+		raw3215_mk_write_req(raw);
+		raw3215_start_io(raw);
+		if ((raw->queued_read || raw->queued_write) &&
+		    !(raw->flags & RAW3215_WORKING) &&
+		    !(raw->flags & RAW3215_TIMER_RUNS)) {
+			raw->timer.expires = RAW3215_TIMEOUT + jiffies;
+			add_timer(&raw->timer);
+			raw->flags |= RAW3215_TIMER_RUNS;
+  		}
 	}
 	spin_unlock_irqrestore(get_ccwdev_lock(raw->cdev), flags);
 }
@@ -318,20 +322,18 @@ static inline void raw3215_try_io(struct
 		    (raw->flags & RAW3215_FLUSHING)) {
 			/* execute write requests bigger than minimum size */
 			raw3215_start_io(raw);
-			if (raw->flags & RAW3215_TIMER_RUNS) {
-				del_timer(&raw->timer);
-				raw->flags &= ~RAW3215_TIMER_RUNS;
-			}
-		} else if (!(raw->flags & RAW3215_TIMER_RUNS)) {
-			/* delay small writes */
-			init_timer(&raw->timer);
-			raw->timer.expires = RAW3215_TIMEOUT + jiffies;
-			raw->timer.data = (unsigned long) raw;
-			raw->timer.function = raw3215_timeout;
-			add_timer(&raw->timer);
-			raw->flags |= RAW3215_TIMER_RUNS;
 		}
 	}
+	if ((raw->queued_read || raw->queued_write) &&
+	    !(raw->flags & RAW3215_WORKING) &&
+	    !(raw->flags & RAW3215_TIMER_RUNS)) {
+		init_timer(&raw->timer);
+		raw->timer.expires = RAW3215_TIMEOUT + jiffies;
+		raw->timer.data = (unsigned long) raw;
+		raw->timer.function = raw3215_timeout;
+		add_timer(&raw->timer);
+		raw->flags |= RAW3215_TIMER_RUNS;
+	}
 }
 
 /*
